### ISSUE TITLE
toIntOption()

### ISSUE THREAD

**Issue by abuffery** [SandroMaglione/fpdart#80]
It would be nice and hopefully easy to add some common helpers like:

```dart
Option<int> toIntOption(String x) => Option.tryCatch(() => int.parse(x));
```

this would allow us to have clean code like:

```dart
assert(Right("1985").map(toIntOption) == Right(Some(1985)));
```

**Comment by SandroMaglione** [comment]
Hi @abuffery 

Do you mean generally a way to parse from one type to another using `fpdart` (`Option`)? Do you have some idea for other helpers or `toIntOption` would be enough?

**Comment by abuffery** [comment]
Hi @SandroMaglione

I am taking as reference Scala that has built-in conversions to Option.

In Scala you can do "41a".toInt and also "41a".toIntOption.

This leads to a leaner syntax, specially when chaining.

For example `"abc".toIntOption.getOrElse(-1)` is easy on the eyes to follow and read.

```scala
"5".toIntOption                 // Option[Int] = Some(5)
"abc".toIntOption               // Option[Int] = None
"abc".toIntOption.getOrElse(-1) // Int = -1
```
I generally find it even clearer when piped on separate lines.

```scala
"abc"
  .toIntOption
  .getOrElse(-1) // Int = -1
```

Other Scala conversions to Option:

- toLongOption
- toDoubleOption 
- toBooleanOption
- toCharOption
- toByteOption 
- toShortOption

In Dart, it would be universally useful for its most basic types:

- toBoolOption
- toDoubleOption
- toIntOption
- toStringOption

Scala does not have a toIntEither and at the same time it may make sense for common basic types.

Currently fpdart has for elaborate cases:

```dart
final tryCatch = Either.tryCatch(
  () => int.parse('invalid'),
  (e, s) => 'Error: $e',
);
```

Why not add to fpdart helpers for basic types when one wants returned a `Left<String>` instead of `None` 

```dart
"5"
  .toIntEither("simple error message")  //  Right(5)

"Not a number"
  .toIntEither("simple error message")  //  Left("simple error message")
```

**Comment by SandroMaglione** [comment]
@abuffery I agree, this helpers may be indeed useful to add to `fpdart`, both for `Option` and `Either`.

I add this to the list of features to add to the library, using this as reference issue for further discussion if necessary 🔜

Thanks for the suggestions 👍
